datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,262 @@
+    /// let expr = col("c2")
+    ///     .lt(lit(1))
+    ///     .or(col("c1").lt(lit(2)))
+    ///     .and(col("c1").lt(lit(3)))
+    ///     .and(lit(false));
+    ///
+    /// // run the rewrite cycle loop
+    /// let (expr, info) = RewriteCycle::new()
+    ///     .with_max_cycles(4)
+    ///     .each_cycle(expr, |cycle_state| {
+    ///         cycle_state
+    ///             .rewrite(&mut const_evaluator)?
+    ///             .rewrite(&mut simplifier)
+    ///     }).unwrap();
+    /// assert_eq!(expr, lit(false));
+    /// assert_eq!(info.completed_cycles(), 2);
+    /// assert_eq!(info.total_iterations(), 4);
+    /// ```
+    ///
+    pub fn each_cycle<
+        Node: TreeNode,
+        F: FnMut(
+            RewriteCycleState<Node>,
+        ) -> RewriteCycleControlFlow<RewriteCycleState<Node>>,
+    >(
+        &self,
+        node: Node,
+        mut f: F,
+    ) -> Result<(Node, RewriteCycleInfo)> {
+        let mut state = RewriteCycleState::new(node);
+        if self.max_cycles == 0 {
+            return state.finish();
+        }
+        // run first cycle then record number of rewriters
+        state = match f(state) {
+            ControlFlow::Break(result) => return result?.finish(),
+            ControlFlow::Continue(node) => node,
+        };
+        state.record_cycle_length();
+        if state.is_done() {
+            return state.finish();
+        }
+        // run remaining cycles
+        match (1..self.max_cycles).try_fold(state, |state, _| f(state)) {
+            ControlFlow::Break(result) => result?.finish(),
+            ControlFlow::Continue(state) => state.finish(),
+        }
+    }
+}
+
+/// Iteration state of a rewrite cycle. See [RewriteCycle::each_cycle] for usage examples and information.
+#[derive(Debug)]
+pub struct RewriteCycleState<Node: TreeNode> {
+    node: Node,
+    consecutive_unchanged_count: usize,
+    rewrite_count: usize,
+    cycle_length: Option<usize>,
+}
+
+impl<Node: TreeNode> RewriteCycleState<Node> {
+    fn new(node: Node) -> Self {
+        Self {
+            node,
+            cycle_length: None,
+            consecutive_unchanged_count: 0,
+            rewrite_count: 0,
+        }
+    }
+
+    /// Records the rewrite cycle length based on the current iteration count
+    ///
+    /// When the total number of writers is not known upfront - such as when using
+    /// [RewriteCycle::each_cycle] we need to keep count of the number of [Self::rewrite]
+    /// calls and then record the number at the end of the first cycle.
+    fn record_cycle_length(&mut self) {
+        self.cycle_length = Some(self.rewrite_count);
+    }
+
+    /// Returns true when the loop has reached the maximum cycle length or when we've received
+    /// consecutive unchanged tree nodes equal to the total number of rewriters.
+    fn is_done(&self) -> bool {
+        // default value indicates we have not completed a cycle
+        let Some(cycle_length) = self.cycle_length else {
```

Comment:
It seems the logic here is because we don't know the number of rules in the first pass.
Is it more readable if we just run the first pass without any checking. And, start to check if early return is possible after the second pass?